src/pytorch_tabular/tabular_model_tuner.py

```diff
@@ -254,27 +254,15 @@ def tune(
         if progress_bar:
             iterator = track(iterator, description=f"[green]{strategy.replace('_',' ').title()}...", total=n_trials)
         verbose_tabular_model = self.tabular_model_init_kwargs.pop("verbose", False)
-        temp_tabular_model = TabularModel(
```

Comment:
Creating the Datamodule inside the loop has some disadvantages. For instance, in case of a large dataset, and you are doing some transformations in the datamodule, it can be a long process. That was why I decided to keep it out of the loop. Maybe we can make it clearer somewhere that parameters in the dataconfig is not tunable.

Reply:
It makes sense. I will think about how to solve this.

Reply:
Changed to sort by trainer_config and just recreate datamodule when trainer_config changes.

If search_space = {
"trainer_config__batch_size": [128, 256],
"optimizer_config__optimizer": ["Adam", "SGD"]
}
Recreate datamodule two times -> [[128, "adam"], [128 "SGD]] - [[256, "adam"], [256 "SGD]] 

If search_space = {
"trainer_config__batch_size": [128, 256],
"trainer_config__max_epochs": [2, 4],
"optimizer_config__optimizer": ["Adam", "SGD"]
}
Recreate datamodule four times -> [[128, 2, "adam"], [128, 2, "SGD"]] - [[128, 4 "adam], [128, 4 "SGD]] - [[256, 2, "adam"], [256, 2, "SGD"]] - [[256, 4 "adam], [256, 4 "SGD]]

If not change trainer_config will create one time!

Reply:
The code is configured to not change datamodule if no changes are made to the TrainerConfig (it will continue training at the same speed as before). I understand that it can take a while to recreate the datamodule when just changing the batch_size (if large dataset) and it is possible to change it without recreate datamodule, but I think it is interesting to support changing the entire TrainerConfig as well for fun experiments (and learning) :D

- Different batch_size improve the results? (https://medium.com/geekculture/how-does-batch-size-impact-your-model-learning-2dd34d9fb1fa)
- Different precisions improve the results? (and lower precision usually uses less memory wich means that it is possible to train if higher batch_size, If result OOM, no problem, ignore that combination e continue)
- Tracking L1 norm improve the results?

If you think it is not useful, I can remove the TrainerConfig and leave just OptimizerConfig and the OOM implemented in this PR, no problem :)

Reply:
All of the TrainerConfig params, except the batch size are set in the PyTorch Trainer. So, really, there is no need to reinitialize datamodule for any of them. batch_size is still set in the datamodule. But as I said before, batch size is only used in the `train_datamodule` methods where the dataloader is created at run time with the batch_size which is set as an attribute of dataloader. So, what we can do is just set the batch_size attribute of the datamodule (which we have already created) and run training with it. Only thing to keep in mind is to set the best batch_size back to the datamodule before returning the best model.

Also, I agree batch size does make a sensible Hyperparameter (although I typically don't tune it that much).  The other ones on your list is a bit weak, but there are some parameters in the trainer config which is worth tuning. So, I have no issue in enabling TrainerConfig params. DataConfig params, probably not. 

So, in summary we can enable tuning for all the TrainerConfig params, even "batch_size" without rebuilding the datamodule.

Reply:
Removed trainer_config from tuner and added an error message if try to use it on the tuner.

Reply:
I didn't mean not to enable tuning for `TrainerConfig`. Apologise for the misunderstanding. My only issue was with re-initializing the datamodule. In the current code also, we pass new `TrainerConfig` each execution of the loop. So, I think we can keep it like that.

the only issue with enabling tuning for `TrainerConfig` is the `batch_size` parameter, when CV is not enabled (correct me if I'm wrong). All the other parameters gets passed to PyTorch Lightning Trainer, which we are re-initializing each iteration of the loop. So, we have two options here:
1. Check specifically for `trainer_config_batch_size` when CV is not enabled and raise an error.
2. Monkeypatch the existing datamodule with new batch size
``` python
datamodule.batch_size = params['batch_size']
```
This should work. Also, just like you are saving best_model to return, also save best batch size. And while returning
``` python
datamodule.batch_size = best_batch size
best_model.datamodule = datamodule
```

Do you want me to try and add commits to this PR?

Reply:
Yes, please. I am a little confused about datamodule :)